upstairs/src/lib.rs

```diff
@@ -1506,56 +1507,140 @@ impl Buffer {
         self.len == 0
     }
 
-    pub async fn as_vec(&self) -> MutexGuard<'_, Vec<u8>> {
-        self.data.lock().await
+    pub fn write(&mut self, offset: usize, data: &[u8]) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            self.data[offset + i] = data[i];
+            self.owned[offset + i] = true;
+        }
     }
 
-    pub async fn owned_vec(&self) -> MutexGuard<'_, Vec<bool>> {
-        self.owned.lock().await
+    pub fn write_with_ownership(
+        &mut self,
+        offset: usize,
+        data: &[u8],
+        owned: &[bool],
+    ) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if owned[i] {
+                self.data[offset + i] = data[i];
+                self.owned[offset + i] = true;
+            }
+        }
+    }
+
+    pub fn write_read_response(
+        &mut self,
+        offset: usize,
+        response: &ReadResponse,
+    ) {
+        assert!(offset + response.data.len() <= self.data.len());
+        if !response.block_contexts.is_empty() {
+            for i in 0..response.data.len() {
+                self.data[offset + i] = response.data[i];
+                self.owned[offset + i] = true;
+            }
+        }
+    }
+
+    pub fn read(&self, offset: usize, data: &mut [u8]) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if self.owned[offset + i] {
+                data[i] = self.data[offset + i];
+            }
+        }
+    }
+
+    pub fn as_bytes(&self) -> Bytes {
```

Comment:
This could be sneakily expensive, since it's doing a full copy.  It could be more efficient if it was consuming (`fn as_bytes(self) -> Bytes`, since `Bytes` implements `From<Vec<u8>>` and reuses the allocation (we'd probably rename as `into_bytes` in that case).

Looking at the code, `Buffer::as_bytes` is only used in two places:
- In `Volume::scrub`, no issues with it being consuming
- In `IOSpan::write_affected_blocks_to_volume`, it _looks_ bad; however, it turns out that we could **also** make `IOSpan::write_affected_blocks_to_volume` consuming.

Reply:
changed in a5ac359